lib/autoPagination.js

```diff
@@ -0,0 +1,190 @@
+'use strict';
+
+var makeRequest = require('./makeRequest');
+var utils = require('./utils');
+
+function getItemCallback(args) {
+  if (args.length === 0) {
+    return undefined;
+  }
+  var onItem = args[0];
+  if (typeof onItem !== 'function') {
+    throw Error('The first argument to autoPagingEach, if present, must be a callback function; receieved ' + typeof onItem);
+  }
+
+  // `.autoPagingEach((item, next) => { doSomething(item); next(); });`
+  if (onItem.length === 2) {
+    return onItem;
+  }
+
+  if (onItem.length > 2) {
+    throw Error('The `onItem` callback function passed to autoPagingEach must accept at most two arguments; got ' + onItem);
+  }
+
+  // API compat; turn this:
+  //    .autoPagingEach((item) => { doSomething(item); return false; });
+  // into this:
+  //    .autoPagingEach((item, next) => { doSomething(item); next(false); });
+  return function(item, next) {
+    var shouldContinue = onItem(item);
+    next(shouldContinue);
+  };
+}
+
+function getDoneCallback(args) {
+  if (args.length < 2) {
+    return undefined;
+  }
+  var onDone = args[1];
+  if (typeof onDone !== 'function') {
+    throw Error('The second argument to autoPagingEach, if present, must be a callback function; receieved ' + typeof onDone);
+  }
+  return onDone;
+}
+
+function wrapAsyncIteratorWithCallback(asyncIteratorNext, onItem) {
+  return new Promise(function(resolve, reject) {
+    function handleIteration(iterResult) {
+      if (iterResult.done) {
+        resolve();
+        return;
+      }
+
+      var item = iterResult.value;
+      return new Promise(function(next) {
+        // Bit confusing, perhaps; we pass a `resolve` fn
+        // to the user, so they can decide when and if to continue.
+        onItem(item, next);
+      }).then(function(shouldContinue) {
+        if (shouldContinue === false) {
+          return handleIteration({done: true});
+        } else {
+          return asyncIteratorNext().then(handleIteration);
+        }
+      });
+    }
+
+    asyncIteratorNext().then(handleIteration).catch(reject);
+  });
+}
+
+function autoPagingEach(self, requestArgs, spec, firstPagePromise) {
```

Comment:
Since these are new methods and fairly non-obvious to figure out how to use based on the function signature, could we try adding some basic documentation blocks for `autoPagingEach` and `autoPagingToArray`?

Reply:
Sure; for users or maintainers? I assume users?